README.md

```diff
@@ -1 +1,122 @@
+
+See the docstring of `populate_result_buffer!` and `consume!` for more information about how to integrate with them.
+
+## Example: Examining the results of the Lexer
+
+Since `ChunkedBase.jl` handles newline detection automatically, a very simple chunked processor could just show the newline positions to the user. Of course, this is not very useful apart from demonstrating how to use this package.
+
+```julia
+using ChunkedBase
+using NewlineLexers: Lexer
+
+# Our chunked processor doesn't require any settings nor does it need to maintain
+# any additional state we'll define our `ParsingContext` and `ConsumeContext` only
+# for dispatch reasons.
+struct ParsingContext <: AbstractParsingContext end
+struct ConsumeContext <: AbstractConsumeContext end
+# Our buffer will hold the newline positions in the current chunk
+struct ResultBuffer <: AbstractResultBuffer
+    newlines_in_chunk::Vector{Int}
+end
+
+# Our overload copies `newlines_segment` to our result buffer
+function ChunkedBase.populate_result_buffer!(
+    result_buf::ResultBuffer,
+    newlines_segment::AbstractVector{Int32},
+    ::ParsingContext,
+    ::Vector{UInt8},
+    ::Union{Nothing,Vector{UInt8}}=nothing,
+    ::Type=Tuple{}
+)
+    resize!(result_buf.newlines_in_chunk, length(newlines_segment))
+    result_buf.newlines_in_chunk .= copy(newlines_segment)
+    return nothing
+end
+
+# We consume result buffer by simply printing it
+function ChunkedBase.consume!(::ConsumeContext, payload::ParsedPayload)
+    chunking_ctx = payload.chunking_ctx
+    result_buffer = payload.results
+
+    # To demonstrate how double-buffering works, print the buffer id and the refill number
+    chunk_info = (chunking_ctx.id, chunking_ctx.buffer_refills[])
+    @info "Newlines in chunk (id:$chunk_info): $(result_buffer.newlines_in_chunk)"
+    return nothing
+end
+```
+
+Now that we defined all overloads and contexts we can define our user-facing function:
+
+```julia
+# Prints relative positions of `\n` in each ingested chunk of size `buffersize`,
+# using `nworkers` spawned tasks.
+function print_newlines(io, buffersize, nworkers)
+    lexer = Lexer(io, nothing, '\n')
+    parsing_ctx = ParsingContext()
+    consume_ctx = ConsumeContext()
+    chunking_ctx = ChunkingContext(buffersize, nworkers, 0, nothing)
+    # ChunkedBase.jl requires 2 result buffers per worker task, we'd get an error otherwise
+    result_buffers = [ResultBuffer(Int[]) for _ in 1:2nworkers]
+    # We need to fill the byte buffer and find the newlines in it before we hand
+    # the `consume_ctx` to the function
+    ChunkedBase.read_and_lex!(lexer, chunking_ctx)
+    parse_file_parallel(lexer, parsing_ctx, consume_ctx, chunking_ctx, result_buffers)
```

Comment:
I don't understand this part of the example. Is it clearer if I go and read the docstrings?

You said earlier that parse_file_parallel is the main entry point. Why do we have to manually invoke a step of the lexer?

Reply:
I'm not sure. .... _Why_ is the expectation that the user has to run one step of the read_and_lex! function?

I think this is the kind of thing I was getting at in our discussion: These kinds of _unexpected_ aspects are what are _most important_ to document clearly. If someone was to come along and see this in RAICode, they might wonder "why are we doing this?" And if they go check the package and it just shows this in the example, but doesn't explain why the code is architected that way, I think it doesn't help a future maintainer who might think they could simplify this and change it, since they wouldn't understand the original motivation for this choice.

I do see now that the docstring says you must do it:
```
    * The `chunking_ctx` is assumed to be filled with data whose newline positions are already detected, e.g.
    by calling `read_and_lex!` with the `lexer` object on it.
```
But I still don't see docs explaining _why_ things are architected this way. I think this is the kind of "understandability" issue that drives the complaints we're trying to resolve here.

Reply:
The reason is that it gives the freedom to implementor packages like `ChunkedCSV` do some schema sniffing and row skipping in the first chunk (or chunks) before we start the actual parsing process. Like if the user wants to find the first row that doesn't start with a comment and is not empty to check if its a header and the skip some more garbage before it actually gets to the parse-able data.

Is this useful context?

Reply:
I've make this step optional and tweaked the docs to reflect that :+1: